File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        int l = 0, r = h.size() - 1;
        int leftmax = 0, rightmax = 0, water = 0;
        while(l < r){
            if(h[l] <= h[r]){
                if(h[l] >= leftmax) leftmax = h[l];
                else water += leftmax - h[l];  
                l++;
            }
            else{
                if(h[r] >= rightmax) rightmax = h[r];    
                else water += rightmax - h[r];
                r--;
            }
        }
        return water;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        hmax, water = 0, 0
        left, right = [], []

        for h in height:
            hmax = max(hmax, h)
            left.append(hmax)

        hmax = 0

        for h in reversed(height):
            hmax = max(hmax, h)
            right.append(hmax)

        right.reverse()

        for h, l, r in zip(height, left, right):
            water += min(l, r) - h

        return water